### PR TITLE
Add global error handler

### PR DESCRIPTION
**Pull request by fetisman**

### CODE REVIEW

File: src/main/java/org/example/advs/controllers/GlobalExceptionController.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+import org.springframework.web.bind.annotation.ExceptionHandler;
+
+@ControllerAdvice
+public class GlobalExceptionController {
+
+    private Logger LOGGER;
+
+
+    @ExceptionHandler(MailException.class)
+    public String handleMailException(MailException e, Model model){
```
Comment (Rib47) [inline src/main/java/org/example/advs/controllers/GlobalExceptionController.java]: Better not to use one-character names for variables or parameters if these are not something like loop index or similar.
Reccomend you to replace "e" with "ex" (at least).

File: src/main/java/org/example/advs/controllers/GlobalExceptionController.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+import javax.mail.AuthenticationFailedException;
+import org.springframework.mail.MailException;
+import org.springframework.ui.Model;
+import org.springframework.web.bind.annotation.ControllerAdvice;
+import org.springframework.web.bind.annotation.ExceptionHandler;
+
+@ControllerAdvice
+public class GlobalExceptionController {
+
+    private Logger LOGGER;
```
Comment (Rib47) [inline src/main/java/org/example/advs/controllers/GlobalExceptionController.java]: Please **always** init logger as a constant for whole class.
Not locally in methods.
 `private static Logger LOGGER = Logger.getLogger("application.properties"); `

File: src/main/java/org/example/advs/controllers/GlobalExceptionController.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+
+    @ExceptionHandler(MailException.class)
+    public String handleMailException(MailException e, Model model){
+        model.addAttribute("mailsendError", "Server Error related to  registration email sending.");
+        LOGGER = Logger.getLogger("application.properties");
+        if (e.getCause() instanceof MailConnectException) { // parent is MailSendException
+                LOGGER.log(Level.WARNING, "MailConnectException. Check out spring.mail.host data in .properties");
```
Comment (Rib47) [inline src/main/java/org/example/advs/controllers/GlobalExceptionController.java]: 1) It is good idea to put more info to your logs for critical situations.
So better to print stack trace in logs:
`LOGGER.log(Level.WARNING, "MailConnectException. Check out spring.mail.host data in .properties", e);`

2) I think that errors descrin=bed in this class worth higher level than "WARNING" ("ERROR").  But only **you** could evaluate importance of some error for your application.

3) The same for the rest of exception types in this class.

File: src/main/java/org/example/advs/controllers/GlobalExceptionController.java [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import javax.mail.AuthenticationFailedException;
+import org.springframework.mail.MailException;
+import org.springframework.ui.Model;
+import org.springframework.web.bind.annotation.ControllerAdvice;
+import org.springframework.web.bind.annotation.ExceptionHandler;
+
+@ControllerAdvice
+public class GlobalExceptionController {
```
Comment (Rib47) [inline src/main/java/org/example/advs/controllers/GlobalExceptionController.java]: 1) I think this class is more like "Handler" than like "Controller".
I would rather put it into "config" package or in the separate package (named "exception" or "error" )

2) **[Global]** Also recommend you to name your **packages** in one style - singular or plural. But singular is more traditional ("controller", "repo", etc., not "repos" or "controller")

File: src/main/java/org/example/advs/controllers/GlobalExceptionController.java [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+    private Logger LOGGER;
+
+
+    @ExceptionHandler(MailException.class)
+    public String handleMailException(MailException e, Model model){
+        model.addAttribute("mailsendError", "Server Error related to  registration email sending.");
+        LOGGER = Logger.getLogger("application.properties");
+        if (e.getCause() instanceof MailConnectException) { // parent is MailSendException
```
Comment (Rib47) [inline src/main/java/org/example/advs/controllers/GlobalExceptionController.java]: Comments looks redundant here.

File: src/main/java/org/example/advs/controllers/GlobalExceptionController.java (hunk 3 above)
Comment (fetisman) [reply src/main/java/org/example/advs/controllers/GlobalExceptionController.java]: Replying on your 2 question , do you mean I have to amend on SEVERE type ?

**Comment by Rib47** [comment]
A few ideas for next PR's
1) "User" already exists after different errors (mail sending, passwords not match). I think you should change operations order: send email before saving user to repository. Or other more complex option (not optimal as for me) - implement "resend registration email" feature for user who tried to register but fail with some errors.

Users are "dead" after these errors - can't login, can't register with this username
Improve error handling.

2) I try set different "password" and "repeat password" on Registration and got no error except "Server Error related to registration email sending." (I will fix this from my side nearest time )) )

**Comment by Rib47** [comment]
4. After sending registr. form I don't see any message like "We just send you activation link, check your email, please"
5. After activation please add motivation text for user
"User successfully activated" -> "User successfully activated. Please enter your user name and password."
7. After click on a button "Create" (registration form) it would be good to make this button inactive at the moment. Sending form takes a time and I think, that nothing happen and click more times on this "Create" button (and got some error).
Possible option to add some styles (CSS) to make button inactive or change color to show to user that you really click is really happen.  If this is possible with FreeMarker
http://htmlbook.ru/css/active
https://www.w3schools.com/howto/howto_css_animate_buttons.asp

**Pull request merged**